ledger/simulation/resources.go

```diff
@@ -605,3 +605,488 @@ func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operatio
 	}
 	return p.tracker.addBox(app, name, readSize, *p.initialBoxSurplusReadBudget, p.ep.Proto.BytesPerBoxReference)
 }
+
+// txnResources tracks the resources being added to a tranasction during resource population
+type txnResources struct {
+	// The static fields are resource arrays that were given in the txn group and thus cannot be removed
+	// The assumption is that these are prefilled because of one of the following reaons:
+	//   - This transaction has already been signed
+	//   - One of the foreign arrays is accessed on-chain
+	staticAssets   map[basics.AssetIndex]struct{}
+	staticApps     map[basics.AppIndex]struct{}
+	staticAccounts map[basics.Address]struct{}
+	staticBoxes    []logic.BoxRef
+
+	// The following fields are fields that are implicitly available to the transaction group from transaction fields
+	assetFromField     basics.AssetIndex
+	appFromField       basics.AppIndex
+	accountsFromFields map[basics.Address]struct{}
+
+	// These are the fields currently being populated, thus we can mutate them however we'd like
+	assets   map[basics.AssetIndex]struct{}
+	apps     map[basics.AppIndex]struct{}
+	boxes    []logic.BoxRef
+	accounts map[basics.Address]struct{}
+
+	maxTotalRefs int
+	maxAccounts  int
+}
+
+func (r *txnResources) getTotalRefs() int {
+	return len(r.accounts) + len(r.assets) + len(r.apps) + len(r.boxes) + len(r.staticAccounts) + len(r.staticAssets) + len(r.staticApps) + len(r.staticBoxes)
+}
+
+// Methods for determining room for specific references
+
+func (r *txnResources) hasRoom() bool {
+	return r.getTotalRefs() < r.maxTotalRefs
+}
+
+func (r *txnResources) hasRoomForAccount() bool {
+	return r.hasRoom() && (len(r.accounts)+len(r.staticAccounts)) < r.maxAccounts
+}
+
+func (r *txnResources) hasRoomForCrossRef() bool {
+	return r.hasRoomForAccount() && r.getTotalRefs() < r.maxTotalRefs-1
+}
+
+func (r *txnResources) hasRoomForBoxWithApp() bool {
+	return r.getTotalRefs() < r.maxTotalRefs-1
+}
+
+// Methods for determining if a resource is available
+
+func (r *txnResources) hasApp(app basics.AppIndex) bool {
+	_, hasStatic := r.staticApps[app]
+	_, hasRef := r.apps[app]
+	return r.appFromField == app || hasStatic || hasRef
+}
+
+func (r *txnResources) hasAsset(aid basics.AssetIndex) bool {
+	_, hasStatic := r.staticAssets[aid]
+	_, hasRef := r.assets[aid]
+	return r.assetFromField == aid || hasStatic || hasRef
+}
+
+func (r *txnResources) hasAccount(addr basics.Address) bool {
+	_, hasStatic := r.staticAccounts[addr]
+	_, hasRef := r.accounts[addr]
+	_, hasField := r.accountsFromFields[addr]
+
+	if r.appFromField.Address() == addr {
+		return true
+	}
+
+	for app := range r.apps {
+		if app.Address() == addr {
+			return true
+		}
+	}
+
+	for app := range r.staticApps {
+		if app.Address() == addr {
+			return true
+		}
+	}
+
+	return hasField || hasStatic || hasRef
+}
+
+func (r *txnResources) addAccount(addr basics.Address) {
```

Comment:
I don't love the style of needing to call `hasRoom()` and then invoking `add*` unconditionally.  Could the `add*` methods return `error` instead?
This would require you have two forms of  some `add*` methods, or an extra argument that means "add this if you can do so by adding only one reference".

Reply:
Let me know what you think of 2ea34f286b9bcea1b7fb2f3c744696bb8e7603c0. Rather than having an additional argument the call sites short circuit if it returns `nil`